### PR TITLE
upstream reviews - take 2

### PR DESCRIPTION
**Pull request by plbossart**
comments on the link DMA support to indicate the current solution is not quite final.
Also fix two issues detected during code reviews (36 patches!), we had 2 issues with error flows for the code loader and the PCM handling. @bardliao you may want to check if the PCM part impacts the load/unload work you are doing.
An additional issue was fixed already by Seppo in a separate PR.

### CODE REVIEW

File: sound/soc/sof/pcm.c [hunk 1]
```diff
@@ -557,12 +556,20 @@ static int sof_pcm_new(struct snd_soc_pcm_runtime *rtd)
...
+		goto free_playback_tables;
 	}
 
 	/* TODO: assign channel maps from topology */
 
+	return ret;
+
+free_playback_tables:
+	if (spcm->pcm.playback)
+		snd_dma_free_pages(&spcm->stream[stream].page_table);
```
Comment (bardliao) [inline sound/soc/sof/pcm.c]: Should it be snd_dma_free_pages(&spcm->stream[ ```SNDRV_PCM_STREAM_PLAYBACK```].page_table);? stream is assigned to SNDRV_PCM_STREAM_CAPTURE after ```capture:``` label.

Comment (keyonjie) [reply sound/soc/sof/pcm.c]: yes, as Bard mentioned.

File: sound/soc/sof/pcm.c [hunk 2]
```diff
@@ -588,6 +595,8 @@ static void sof_pcm_free(struct snd_pcm *pcm)
 	if (spcm->pcm.capture)
 		snd_dma_free_pages(&spcm->stream[SNDRV_PCM_STREAM_CAPTURE].page_table);
 
+	snd_pcm_lib_preallocate_free_for_all(pcm);
+
```
Comment (keyonjie) [inline sound/soc/sof/pcm.c]: we can add this from to private_free() here, though it is not essential as we have another chance to free it at snd_pcm_free() if we don't do that in sof_pcm_free()/private_free().

File: sound/soc/sof/pcm.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/pcm.c]: right, I did a last minute optimization and remove the assignment of the stream variable.
Thanks for the review, will resubmit.

File: sound/soc/sof/pcm.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/pcm.c]: sorry @keyonjie I did not understand your suggestion. I have no idea what private_free refers to in this context.

Comment (bardliao) [reply sound/soc/sof/pcm.c]: @plbossart snd_pcm_lib_preallocate_free_for_all() is already called by snd_pcm_free() (sound/core/pcm.c). So, we don't need to call it in our driver.

Comment (plbossart) [reply sound/soc/sof/pcm.c]: It's a valid point. I was looking at the example in the Skylake driver but it seems to be one of the few exceptions. I'll resubmit and add a comment that the pages are expected to be freed by the core, that way the intent is clearly documented.

````text
git grep snd_pcm_lib_preallocate_free
core/pcm.c:     snd_pcm_lib_preallocate_free_for_all(pcm);
core/pcm_memory.c: * snd_pcm_lib_preallocate_free - release the preallocated buffer of the specified substream.
core/pcm_memory.c:int snd_pcm_lib_preallocate_free(struct snd_pcm_substream *substream)
core/pcm_memory.c: * snd_pcm_lib_preallocate_free_for_all - release all pre-allocated buffers on the pcm
core/pcm_memory.c:int snd_pcm_lib_preallocate_free_for_all(struct snd_pcm *pcm)
core/pcm_memory.c:                      snd_pcm_lib_preallocate_free(substream);
core/pcm_memory.c:EXPORT_SYMBOL(snd_pcm_lib_preallocate_free_for_all);
soc/dwc/dwc-pcm.c:      snd_pcm_lib_preallocate_free_for_all(pcm);
soc/intel/skylake/skl-pcm.c:    snd_pcm_lib_preallocate_free_for_all(pcm);
soc/mediatek/common/mtk-afe-platform-driver.c:  snd_pcm_lib_preallocate_free_for_all(pcm);
soc/stm/stm32_adfsdm.c:         snd_pcm_lib_preallocate_free_for_all(pcm);
soc/uniphier/aio-dma.c: snd_pcm_lib_preallocate_free_for_all(pcm);
````

**Comment by plbossart** [comment]
@RanderWang we can do whatever changes we want to sof structures since we 'own' the framework. Modifying or abusing ALSA or ASoC APIs and structures is not ok in general, we are not the owners and need to have a strong case to get a change accepted by Mark Brown and Takashi Iwai.

**Pull request merged**